### PR TITLE
Test hedging hysteresis

### PR DESCRIPTION
**Pull request by sebastienverreault**
add hedging hysteresis unit tests
fix minimum liability threshold 
add minimum liability threshold unit tests
add dealer v1 scenario based integration test

### CODE REVIEW

File: hedging/tests/fixtures/hedging.json [hunk 1]
```diff
@@ -10,4 +10,4 @@
       "liability": "0"
     }
   ]
-}
+}
```
Comment (bodymindarts) [inline hedging/tests/fixtures/hedging.json:13]: You need to fix your editor to add new lines at end of file.

File: hedging/tests/fixtures/scenario_01.json [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+            "liability": "55"
+        },
+        {
+            "liability": "30"
+        },
+        {
+            "liability": "15"
+        }
+    ]
+}
```
Comment (bodymindarts) [inline hedging/tests/fixtures/scenario_01.json:40]: Missing newline.

File: hedging/tests/hedging.rs [hunk 3]
```diff
@@ -138,3 +139,69 @@ async fn hedging() -> anyhow::Result<()> {
...
+                okex_poll_frequency: std::time::Duration::from_secs(1),
+            },
+            okex_client_config(),
+            pubsub_config,
+        )
+        .await
+    });
+    tokio::time::sleep(std::time::Duration::from_secs(3)).await;
+
+    let expected_exposure = [
```
Comment (bodymindarts) [inline hedging/tests/hedging.rs:177]: I don't like that this is defined here while the fixtures are in the json blob.
Can we either derive the expectation from the payload or encode them together?
Perhaps:
scenario_01.json
```
{
  scenario: [
  { payload: { liability: 100}
    expectation: { lowerBound: 99, upperBound: 101 }
    },
    <etc>
  ]
}
```

File: hedging/src/adjustment_action.rs [hunk 4]
```diff
@@ -164,10 +164,97 @@ mod tests {
...
+        let liability = Decimal::from(nominal_liability);
+        let exposure = Decimal::from(-nominal_liability) * LOW_BOUND_RATIO_SHORTING;
+
+        let adjustment = determine_action(liability, exposure);
+        assert_eq!(adjustment, AdjustmentAction::DoNothing);
+    }
+
+    #[test]
+    fn low_bound_below() {
+        // condition: exposure / liability < LOW_BOUND_RATIO_SHORTING
```
Comment (bodymindarts) [inline hedging/src/adjustment_action.rs:187]: It seems like these tests are simply mirroring the implementation... hmm.

Also I'm not a fan of the comments. I see their value but perhaps they should at least be moved above the fn as doc comments:
https://doc.rust-lang.org/rust-by-example/meta/doc.html
```
/// ...
#[test]
fn ..
```

File: hedging/tests/hedging.rs [hunk 5]
```diff
@@ -114,7 +112,10 @@ async fn hedging() -> anyhow::Result<()> {
     });
     tokio::time::sleep(std::time::Duration::from_secs(3)).await;
 
-    let mut payloads = load_fixture()?.payloads.into_iter();
+    let mut payloads = load_fixture("./tests/fixtures/hedging.json")
```
Comment (bodymindarts) [inline hedging/tests/hedging.rs:115]: I think we can delete the `hedging` test - `scenario_01` seems to capture everything - so its redundant.

File: hedging/src/adjustment_action.rs (hunk 4 above)
Comment (sebastienverreault) [reply hedging/src/adjustment_action.rs:187]: I can do the calculations offline and put magic numbers here, it will be the same "mirroring" of the implementation, no choice else they won't pass. So those unit test will protect from implementation changes breaking them.

Comments were for me before writing the code, I can remove them, we don't have any, anywhere.

What do you think?

File: hedging/tests/hedging.rs (hunk 5 above)
Comment (sebastienverreault) [reply hedging/tests/hedging.rs:115]: Ok, removing.

File: hedging/tests/fixtures/hedging.json (hunk 1 above)
Comment (sebastienverreault) [reply hedging/tests/fixtures/hedging.json:13]: I see, I was wondering about this... thx

File: hedging/tests/hedging.rs (hunk 3 above)
Comment (sebastienverreault) [reply hedging/tests/hedging.rs:115]: Yeah wanted to do exactly that, and was going to ask what you thought was the best way...

Comment (bodymindarts) [reply hedging/tests/hedging.rs:115]: I think the best way would be to derive the expectation from the `payload.liability`. If the upper / lower bound is always +/- 1 this should be simple.

File: hedging/src/adjustment_action.rs (hunk 4 above)
Comment (bodymindarts) [reply hedging/src/adjustment_action.rs:187]: If you don't need them anymore my preference is to delete the comments.

no magic numbers - I guess the mirroring is okay for now - can't think of anything better.

File: hedging/tests/fixtures/scenario_01.json [hunk 6]
```diff
@@ -0,0 +1,40 @@
+{
+    "payloads": [
```
Comment (bodymindarts) [inline hedging/tests/fixtures/scenario_01.json:2]: I don't really understand this test. What edge cases / code paths is it going through that the `hedging` test didn't touch?
All things being equal we should prefer the test with the shortest runtime. I'm wondering if this is actually redundant and we keep the original one - or perhaps extend it slightly if there is a code path we missed.

Note that integration tests are not intended to test the exact logic - that's what unit tests are for.

Comment (sebastienverreault) [reply hedging/tests/fixtures/scenario_01.json:2]: All right, in that case I will restore your test and remove this one and we should be good to go.

File: hedging/src/adjustment_action.rs (hunk 4 above)
Comment (sebastienverreault) [reply hedging/src/adjustment_action.rs:187]: perfect will do that. thx

**Comment by bodymindarts** [comment]
merged locally
